src/puppetlabs/puppetdb/pql/transform.clj

```diff
@@ -0,0 +1,123 @@
+(ns puppetlabs.puppetdb.pql.transform
+  (:require [clojure.string :as str]
+            [puppetlabs.puppetdb.cheshire :as json]))
+
+(defn transform-from
+  ([entity]
+   ["from" entity])
+  ([entity arg2]
+   ["from" entity arg2])
+  ([entity extract expression]
+   ["from" entity (apply vector (concat extract [expression]))])
+  ([entity extract expression modifier]
+   ["from" entity (apply vector (concat extract [expression modifier]))]))
+
+(defn transform-subquery
+  ([entity]
+   ["subquery" entity])
+  ([entity arg2]
+   ["subquery" entity arg2]))
+
+(defn transform-extract
+  [& args]
+  ["extract" (apply vector args)])
+
+(defn transform-expr4
```

Comment:
So, I did have this once before, and I can go back to it ... I'd also want to update the EBNF to reflect that also. +1.